internal/webhooks/cluster_test.go

```diff
@@ -1833,6 +1833,112 @@ func TestMovingBetweenManagedAndUnmanaged(t *testing.T) {
+		Name:       "baz",
+		Namespace:  "default",
+	}
+
+	topology := builder.ClusterTopology().WithClass("class1").WithVersion("v1.24.3").Build()
+	notFoundTopology := builder.ClusterTopology().WithClass("class2").WithVersion("v1.24.3").Build()
+	baseClusterClass := builder.ClusterClass(metav1.NamespaceDefault, "class1").
+		WithInfrastructureClusterTemplate(refToUnstructured(ref)).
+		WithControlPlaneTemplate(refToUnstructured(ref)).
+		WithControlPlaneInfrastructureMachineTemplate(refToUnstructured(ref))
+
+	// ccFullyReconciled is a ClusterClass with a matching generation and observed generation, and VariablesReconciled=true.
+	ccFullyReconciled := baseClusterClass.DeepCopy().Build()
+	ccFullyReconciled.Generation = 1
+	ccFullyReconciled.Status.ObservedGeneration = 1
+	conditions.MarkTrue(ccFullyReconciled, clusterv1.ClusterClassVariablesReconciledCondition)
+
+	// ccGenerationMismatch is a ClusterClass with a mismatched generation and observed generation, but a true VariablesReconciledCondition.
+	ccGenerationMismatch := baseClusterClass.DeepCopy().Build()
+	ccGenerationMismatch.Generation = 999
+	ccGenerationMismatch.Status.ObservedGeneration = 1
+	conditions.MarkTrue(ccGenerationMismatch, clusterv1.ClusterClassVariablesReconciledCondition)
+
+	// ccGenerationMismatch is a ClusterClass with a mismatched generation and observed generation, but a true VariablesReconciledCondition.
+	ccVariablesReconciledFalse := baseClusterClass.DeepCopy().Build()
+	conditions.MarkFalse(ccGenerationMismatch, clusterv1.ClusterClassVariablesReconciledCondition, "", clusterv1.ConditionSeverityError, "")
+
+	tests := []struct {
+		name         string
+		cluster      *clusterv1.Cluster
+		oldCluster   *clusterv1.Cluster
+		clusterClass *clusterv1.ClusterClass
+		wantErr      bool
+	}{
+		{
+			name:         "Pass on create if ClusterClass is fully reconciled'",
+			cluster:      builder.Cluster(metav1.NamespaceDefault, "cluster1").WithTopology(topology).Build(),
+			clusterClass: ccFullyReconciled,
+			wantErr:      false,
+		},
+		{
+			name:         "Pass on create if ClusterClass generation does not match observed generation'",
+			cluster:      builder.Cluster(metav1.NamespaceDefault, "cluster1").WithTopology(topology).Build(),
+			clusterClass: ccGenerationMismatch,
+			wantErr:      false,
+		},
+		{
+			name:         "Pass on create if ClusterClass generation has False for VariablesReconciled'",
+			cluster:      builder.Cluster(metav1.NamespaceDefault, "cluster1").WithTopology(topology).Build(),
+			clusterClass: ccVariablesReconciledFalse,
+			wantErr:      false,
+		},
+		{
+			name:         "Pass on create if ClusterClass is not found'",
+			cluster:      builder.Cluster(metav1.NamespaceDefault, "cluster1").WithTopology(notFoundTopology).Build(),
+			clusterClass: ccVariablesReconciledFalse,
+			wantErr:      false,
+		},
+		{
+			name:         "Pass on update if oldCluster ClusterClass generation does not match observed generation'",
```

Comment:
This passes because the class is the same for old and new Cluster and we don't call pollClusterClassForCluster for old cluster.

Looking at the code we would fail if on update the oldCluster ClusterClass has a generation mismatch

Should we split this into two cases?

Reply:
Added a pass case and a fail case for update here.